src/uct/ib/mlx5/gdaki/gdaki.cuh

```diff
@@ -93,29 +93,99 @@ template<ucs_device_level_t level> UCS_F_DEVICE void uct_rc_mlx5_gda_sync(void)
     }
 }
 
-UCS_F_DEVICE uint64_t
-uct_rc_mlx5_gda_reserv_wqe_thread(uct_rc_gdaki_dev_ep_t *ep, unsigned count)
+UCS_F_DEVICE uint16_t uct_rc_mlx5_gda_bswap16(uint16_t x)
 {
-    /* Try to reserve optimistically */
-    int32_t prev = atomicAdd(&ep->avail_count, -(int32_t)count);
-    if (prev < (int32_t)count) {
-        /* Rollback */
-        atomicAdd(&ep->avail_count, count);
+    uint32_t ret;
+    asm volatile("{\n\t"
+                 ".reg .b32 mask;\n\t"
+                 ".reg .b32 ign;\n\t"
+                 "mov.b32 mask, 0x1;\n\t"
+                 "prmt.b32 %0, %1, ign, mask;\n\t"
+                 "}"
+                 : "=r"(ret)
+                 : "r"((uint32_t)x));
+    return ret;
+}
+
+UCS_F_DEVICE uint64_t uct_rc_mlx5_gda_parse_cqe(uct_rc_gdaki_dev_ep_t *ep,
+                                                uint16_t *wqe_cnt,
+                                                uint8_t *opcode)
+{
+    auto *cqe64        = reinterpret_cast<mlx5_cqe64*>(ep->cqe_daddr);
+    uint32_t *data_ptr = (uint32_t*)&cqe64->wqe_counter;
+    uint32_t data      = READ_ONCE(*data_ptr);
+    uint64_t rsvd_idx  = READ_ONCE(ep->sq_rsvd_index);
+
+    *wqe_cnt = uct_rc_mlx5_gda_bswap16(data);
+    if (opcode != nullptr) {
+        *opcode = data >> 28;
+    }
+
+    return rsvd_idx - ((rsvd_idx - *wqe_cnt) & 0xffff);
```

Comment:
Maybe better use 0xffffull?

Reply:
smaller values always promoted to bigger so it's automatically `ull`